stdlib/private/PythonUnittest/PythonUnittest.swift

```diff
@@ -0,0 +1,358 @@
+import Python
+import StdlibUnittest
+
+/// The gc module is an interface to the Python garbage collector.
+let gc = Python.import("gc")
+/// The tracemalloc module is a debug tool to trace memory blocks allocated by
+/// Python. It is optionally imported because it is available only in Python 3.
+let tracemalloc = try? Python.attemptImport("tracemalloc")
+
+extension TestSuite {
+  /// Check that running `body` does not cause Python memory leaks.
+  public func testWithLeakChecking(_ name: String, body: @escaping () -> Void) {
+    test(name) {
+      if let tracemalloc = tracemalloc {
+        tracemalloc.start()
+      }
+      // Note: Convert to integer to prevent the integer from messing up
+      // tracemalloc's count.
+      let referencedObjectCount = Int(Python.len(gc.get_objects()))!
+      body()
+      expectEqual(referencedObjectCount, Int(Python.len(gc.get_objects()))!,
+                  "Python memory leak.")
+      if let tracemalloc = tracemalloc {
+        expectEqual(0, Int(tracemalloc.get_traced_memory().tuple2.0)!)
+        tracemalloc.stop()
+      }
+    }
+  }
+}
+
+extension TestSuite {
+  public func addTestsForVersion(_ major: Int, _ minor: Int) {
+    
+    let versionPrefix = "\(major).\(minor)."
+
+    self.testWithLeakChecking("\(versionPrefix)CheckVersion") {
```

Comment:
Why is it necessary to move tests to stdlib? That kind of makes these tests not tests anymore.

I was expecting a lightweight test utility that takes a bunch of versions that individual test cases can specify, not moving all tests elsewhere. Something like:

```swift
PythonRuntimeTestSuite.testWithLeakChecking("Range", versions: (2, 7), (3, 6)) {
  ...
}
```

Reply:
I believe specifying Python-versions-to-test as an argument (e.g. `versions: (2, 7), (3, 6)`) is difficult to support, since dynamically switching Python version is not supported.

Please check out https://github.com/apple/swift/pull/23745#issuecomment-479246345.

Reply:
I see. Thanks for the context. Moving tests to stdlib/private/ is definitely non-standard, and I like the option of using multiple RUN-lines. Do we have a JIRA issue about dynamic version switching?

Reply:
After [some discussion](https://github.com/apple/swift/pull/23055#issuecomment-469418796), dynamically switching Python versions has been deemed a non-goal.

Pasting relevant context:

> > @pvieito: could you please confirm the behavior of `PythonLibrary.useVersion`? What are your thoughts on dynamically switching Python versions?
> 
> Loading two versions of `Python` in the same process is not possible, as a lot of common symbols would clash.
> 
> Switching Python versions (in the sense of unloading a Python library and loading other version) could be _technically possible_ in some OS (`dlclose`?), but would make the implementation a lot more complex and unstable (what happens when unloading Python 3 while holding a Python 3 `str` reference and then passing it to a Python 2 function?).

---

Filed [TF-415](https://bugs.swift.org/browse/TF-415) and closed it as `Won't fix`.

Reply:
Yeah, symbol collision is a hard problem. However, if we are sure dynamic switching is not supported, `PythonLibrary.useVersion` should print an error when a version is already selected.

Reply:
> However, if we are sure dynamic switching is not supported, `PythonLibrary.useVersion` should result in a precondition failure when a version is already selected.

That's a great suggestion! Let me file a bug.

Reply:
@dan-zheng thanks for the explanations you gave before on this.  I'll add my comments on the other conversation you started.

and indeed ... TF-416 would make `PythonLibrary.useVersion` more robust.